File: tests/axis_runtime/integration/python/multiple_results_python_2/multiple_results_python_2_app/axis_packages/exaxission/default_exaxission_python/exaxission.py
#
# Copyright © 2025 Agora
# This file is part of APTIMA Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
from typing import Optional

from aptima import Cmd, CmdResult, Extension, StatusCode, TenEnv, TenError


class DefaultExtension(Extension):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        self.name = name

        self.__counter = 0

    def on_init(self, aptima_env: TenEnv) -> None:
        aptima_env.log_debug("on_init")
        aptima_env.on_init_done()

    def check_hello(
        self,
        aptima_env: TenEnv,
        result: Optional[CmdResult],
        error: Optional[TenError],
        receivedCmd: Cmd,
    ):
        if error is not None:
            assert False, error

        assert result is not None

        self.__counter += 1

        if self.__counter == 1:
            assert result.is_completed() is True
            aptima_env.log_info("receive 1 cmd result")

            respCmd = CmdResult.create(StatusCode.OK)
            respCmd.set_property_string("detail", "nbnb")
            aptima_env.return_result(respCmd, receivedCmd)
        else:
            assert False

    def on_cmd(self, aptima_env: TenEnv, cmd: Cmd) -> None:
        cmd_json = cmd.get_property_to_json()
        aptima_env.log_debug(f"on_cmd json: {cmd_json}")

        if self.name == "default_extension_python_1":
            new_cmd = Cmd.create("hello")
            aptima_env.send_cmd(
                new_cmd,
                lambda aptima_env, result, error: self.check_hello(
                    aptima_env, result, error, cmd
                ),
            )
        elif self.name == "default_extension_python_2":
            aptima_env.log_info("create respCmd")
            respCmd = CmdResult.create(StatusCode.OK)
            aptima_env.return_result(respCmd, cmd)
        elif self.name == "default_extension_python_3":
            aptima_env.log_info("create respCmd")
            respCmd = CmdResult.create(StatusCode.OK)
            aptima_env.return_result(respCmd, cmd)

File: tests/axis_runtime/integration/go/prepare_to_stop_go/prepare_to_stop_go_app/axis_packages/exaxission/exaxission_a/exaxission.go
// Copyright © 2025 Agora
// This file is part of APTIMA Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to https://github.com/aptima-ai/aptima_framework/LICENSE for more
// information.
//
// Note that this is just an example extension written in the GO programming
// language, so the package name does not equal to the containing directory
// name. However, it is not common in Go.
package default_extension_go

import (
	"fmt"

	"aptima_framework/aptima"
)

type aExtension struct {
	name      string
	isStopped bool
	aptima.DefaultExtension
}

func newAExtension(name string) aptima.Extension {
	return &aExtension{name: name, isStopped: false}
}

func (p *aExtension) OnDeinit(tenEnv aptima.TenEnv) {
	defer tenEnv.OnDeinitDone()

	tenEnv.LogDebug("onDeinit")
	if !p.isStopped {
		panic("should not happen.")
	}
}

func (p *aExtension) OnCmd(
	tenEnv aptima.TenEnv,
	cmd aptima.Cmd,
) {
	go func() {
		cmdName, _ := cmd.GetName()
		tenEnv.LogInfo(
			"receive command: " + cmdName,
		)
		if cmdName == "start" {
			tenEnv.SendCmd(cmd, func(r aptima.TenEnv, cs aptima.CmdResult, e error) {
				r.ReturnResultDirectly(cs, nil)
			})
		}
	}()
}

func (p *aExtension) OnStop(tenEnv aptima.TenEnv) {
	go func() {
		tenEnv.LogDebug("onStop ")

		cmd, _ := aptima.NewCmd("stop")
		respChan := make(chan aptima.CmdResult, 1)

		tenEnv.SendCmd(
			cmd,
			func(tenEnv aptima.TenEnv, cmdResult aptima.CmdResult, e error) {
				respChan <- cmdResult
			},
		)

		select {
		case resp := <-respChan:
			statusCode, _ := resp.GetStatusCode()
			if statusCode == aptima.StatusCodeOk {
				p.isStopped = true
				tenEnv.OnStopDone()
			} else {
				panic("stop failed.")
			}
		}
	}()
}

func init() {
	fmt.Println("call init")

	// Register addon
	err := aptima.RegisterAddonAsExtension(
		"extension_a",
		aptima.NewDefaultExtensionAddon(newAExtension),
	)
	if err != nil {
		fmt.Println("register addon failed", err)
	}
}
